feat(cli): add auth command (#11004)

elasticspoon opened the pull request:
Fixes #11004 (partially)

Creates the command `coder auth` for authentication purposes.

`coder auth` has 3 sub-commands `login`, `status` and `token`. Currently the sub-commands make 2 requests to the server: validation of the user and an additional command for sub-command info (token info, etc.)

`coder auth status` and `coder auth token` commands return an error if the user is not authenticated when they are called.

```console
$ > coder auth status
Encountered an error running "coder auth status", see "coder auth status --help" for more information
error: You are not logged in. Try logging in using 'coder login <url>'.
exit status 1
```

### `coder auth`

Just returns the help message.

### `coder auth login`

This is the same command as `coder login`.

### `coder auth status`

Informs the user about their authentication status. It may make sense for this to not show an error if called when user is not logged in.

```console
$ > coder auth status
Hello there, admin! You're authenticated at http://localhost:8080.
```

Additional info that is available from the current requests and could be shown:

- ID
- Username
- AvatarURL
- Name
- Email
- CreatedAt
- LastSeenAt
- Status
- LoginType
- ThemePreference
- OrganizationIDs
- Roles

### `coder auth token`

Informs the user of their token string and the expiration time of the token. (Maybe `coder auth session` is a better name or alias?)

```console
$ > coder auth token
Your session token 'DqSDeIXhgx-ABpLuIJ5PnLJjtvdUwmRK5' expires at 2024-04-13 18:08:59.278587 -0400 EDT.
```

Additional info that is available from the current requests and could be shown:

- ID
- UserID
- LastUsed
- ExpiresAt
- CreatedAt
- UpdatedAt
- LoginType
- Scope
- TokenName
- LifetimeSeconds

elasticspoon committed:
feat(cli): add auth command (#11004)

Adds `coder auth` command with 3 sub-commands: `coder auth token`,
`coder auth status` and `coder auth login`.

`coder auth login`: This is the same command as `coder login`.

`coder auth status`: Informs the user about their authentication
status. It may make sense for this to not show an error
if called when user is not logged in.

`coder auth token`: Informs the user of their token string
and the expiration time of the token.

All the commands use a middleware to check that the user is
authenticated and will return an error if the user is not. They also
make at most 2 API calls.

The commands have additional values that could be shown but are not with
this implementation.

mafredri commented on cli/auth.go:
```diff
@@ -0,0 +1,87 @@
+				return err
+			}
+
+			_, _ = fmt.Fprintf(inv.Stdout, "Your session token '%s' expires at %s.\n", client.SessionToken(), key.ExpiresAt)
```
Suggestion: Could be nice to show a humanized relative time, like `(in 6.8 days)`.

(Don't worry, not necessary for this PR.)

mafredri commented on cli/auth.go:
```diff
@@ -0,0 +1,87 @@
+		return func(inv *serpent.Invocation) error {
+			_, err := client.User(inv.Context(), codersdk.Me)
+			if err != nil {
+				return xerrors.Errorf("get user: %w", err)
```
I feel the `get user` wrapping here adds a bit of indirection and makes the error harder to understand:

```
❯ ./coder auth token --token foo
Encountered an error running "coder auth token", see "coder auth token --help" for more information
error: Trace=[get user: ]
You are signed out or your session has expired. Please sign in again to continue.
Suggestion: Try logging in using 'coder login'.
```

Although, since this is only used once, how about moving the logic to `authToken` handler instead? Is there a benefit of it being a middleware?

elasticspoon replied:
You right. That was some premature optimization, I thought I would be using this more but then I made changes to reduce DB calls and left this in.

mafredri commented on cli/auth.go:
```diff
@@ -0,0 +1,87 @@
+func (r *RootCmd) auth() *serpent.Command {
+	cmd := &serpent.Command{
+		Use:   "auth <subcommand>",
+		Short: "Manage information about internal authentication.",
```
Maybe something like:

```suggestion
		Short: "Authenticate with a Coder deployment and view authentication status.",
```

Or:

```suggestion
		Short: "Manage authentication for Coder deployment.",
```

I feel "internal" here could be confusing to users.

mafredri commented on cli/auth.go:
```diff
@@ -0,0 +1,87 @@
+	client := new(codersdk.Client)
+	cmd := &serpent.Command{
+		Use:   "token",
+		Short: "Show session token value and expiration time.",
```
```suggestion
		Short: "Show the current session token and expiration time.",
```

mafredri reviewed (commented):
I noticed when the user provides `--token` or the env, this might lead to some confusion as to where auth status originates from. It might be good to show "source" here, i.e. either it's from config, env or flag.

elasticspoon committed:
fixup!: review changes

Change language in commands
Humanize session token remaining time
Inline auth middleware

elasticspoon commented:
@matifali @mafredri Missed that this got reviewed, addressed both reviews.
